File: src/lib/api/shell.ts
// F006 — Tauri shell polish / settings client.
//
// Backs the AppShellSettings + ProcessHealthIndicator components. All calls
// route through the sidecar's /shell/* endpoints. Tauri-command-level shell
// integration (sidecar_port, processes, restart_sidecar) lives in
// ../../features/shell/* and falls back to these HTTP endpoints when the
// Tauri command bridge is not available (e.g. in plain `vite dev`).

import { getJSON, postJSON } from "../api";

export interface ManagedProcess {
  pid: number;
  name: string;
  role: string;
  status: string;
  cpu_percent: number;
  rss_bytes: number;
  started_at: number | null;
}

export interface ProcessesResponse {
  processes: ManagedProcess[];
}

export interface OllamaHealth {
  host: string;
  reachable: boolean;
  version?: string | null;
  error?: string | null;
}

export interface SidecarSubHealth {
  pid: number;
  uptime_seconds: number;
  cold_start_seconds: number | null;
}

export interface ShellStatus {
  sidecar: SidecarSubHealth;
  ollama: OllamaHealth;
  processes: ManagedProcess[];
}

export interface SidecarPortInfo {
  port: number;
  source: "env" | "default";
}

export interface OllamaHostInfo {
  host: string;
}

export interface ColdStartInfo {
  cold_start_seconds: number | null;
  process_start_epoch: number;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function numberOr(value: unknown, fallback: number): number {
  return typeof value === "number" && Number.isFinite(value) ? value : fallback;
}

function nullableNumber(value: unknown): number | null {
  return typeof value === "number" && Number.isFinite(value) ? value : null;
}

function stringOrNull(value: unknown): string | null {
  return typeof value === "string" ? value : null;
}

function stringOr(value: unknown, fallback: string): string {
  return typeof value === "string" ? value : fallback;
}

function normalizeManagedProcess(value: unknown): ManagedProcess | null {
  if (!isRecord(value)) return null;
  const pid = numberOr(value.pid, NaN);
  if (!Number.isFinite(pid)) return null;
  return {
    pid,
    name: stringOr(value.name, "unknown"),
    role: stringOr(value.role, "unknown"),
    status: stringOr(value.status, "unknown"),
    cpu_percent: numberOr(value.cpu_percent, 0),
    rss_bytes: numberOr(value.rss_bytes, 0),
    started_at: nullableNumber(value.started_at),
  };
}

export function normalizeProcessesResponse(raw: unknown): ProcessesResponse {
  const processes = isRecord(raw) && Array.isArray(raw.processes)
    ? raw.processes
        .map((p) => normalizeManagedProcess(p))
        .filter((p): p is ManagedProcess => p !== null)
    : [];
  return { processes };
}

export function normalizeShellStatus(raw: unknown): ShellStatus {
  const root = isRecord(raw) ? raw : {};
  const sidecar = isRecord(root.sidecar) ? root.sidecar : {};
  const ollama = isRecord(root.ollama) ? root.ollama : {};
  const reachable = ollama.reachable === true;

  return {
    sidecar: {
      pid: numberOr(sidecar.pid, 0),
      uptime_seconds: numberOr(sidecar.uptime_seconds, 0),
      cold_start_seconds: nullableNumber(sidecar.cold_start_seconds),
    },
    ollama: {
      host: stringOr(ollama.host, ""),
      reachable,
      version: stringOrNull(ollama.version),
      error: stringOrNull(ollama.error) ?? (reachable ? null : "status unavailable"),
    },
    processes: normalizeProcessesResponse(root).processes,
  };
}

export function normalizeSidecarPort(raw: unknown): SidecarPortInfo {
  const root = isRecord(raw) ? raw : {};
  const port = numberOr(root.port, 8770);
  return {
    port: port > 0 && port <= 65535 ? port : 8770,
    source: root.source === "env" ? "env" : "default",
  };
}

export function status(): Promise<ShellStatus> {
  return getJSON<unknown>("/shell/status").then(normalizeShellStatus);
}

export function processes(): Promise<ProcessesResponse> {
  return getJSON<unknown>("/shell/processes").then(normalizeProcessesResponse);
}

export function sidecarPort(): Promise<SidecarPortInfo> {
  return getJSON<unknown>("/shell/sidecar/port").then(normalizeSidecarPort);
}

export function getOllamaHost(): Promise<OllamaHostInfo> {
  return getJSON<OllamaHostInfo>("/shell/config/ollama-host");
}

export function setOllamaHost(host: string): Promise<OllamaHostInfo> {
  return postJSON<OllamaHostInfo>("/shell/config/ollama-host", { host });
}

export function markReady(): Promise<ColdStartInfo> {
  return postJSON<ColdStartInfo>("/shell/ready");
}

export function coldStart(): Promise<ColdStartInfo> {
  return getJSON<ColdStartInfo>("/shell/cold-start");
}

File: python/errorta_app/routes/shell.py
"""F006 — Tauri shell polish / settings router.

Exposes:
  GET  /shell/status               — sidecar + Ollama health roll-up
  GET  /shell/processes            — list managed child PIDs (sidecar + Ollama)
  GET  /shell/sidecar/port         — return the port the sidecar is bound to
  GET  /shell/config/ollama-host   — current Ollama host (shell-tier override)
  POST /shell/config/ollama-host   — update Ollama host (forwards to F003 later)
  POST /shell/ready                — frontend ping that records cold-start time
  GET  /shell/cold-start           — cold-start measurement (seconds)
"""
from __future__ import annotations

import os
from typing import Optional

import httpx
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from errorta_shell import config as shell_config
from errorta_shell import processes as shell_processes

router = APIRouter(prefix="/shell", tags=["shell"])


# --- Models -----------------------------------------------------------------


class ProcessPayload(BaseModel):
    pid: int
    name: str
    role: str
    status: str
    cpu_percent: float
    rss_bytes: int
    started_at: Optional[float] = None


class ProcessesResponse(BaseModel):
    processes: list[ProcessPayload]


class OllamaHealth(BaseModel):
    host: str
    reachable: bool
    version: Optional[str] = None
    error: Optional[str] = None


class SidecarHealth(BaseModel):
    pid: int
    uptime_seconds: float
    cold_start_seconds: Optional[float] = None


class StatusResponse(BaseModel):
    sidecar: SidecarHealth
    ollama: OllamaHealth
    processes: list[ProcessPayload]


class SidecarPortResponse(BaseModel):
    port: int
    source: str  # "env" | "default"


class OllamaHostBody(BaseModel):
    host: str = Field(min_length=1)


class OllamaHostResponse(BaseModel):
    host: str


class ColdStartResponse(BaseModel):
    cold_start_seconds: Optional[float] = None
    process_start_epoch: float


# --- Helpers ----------------------------------------------------------------


def _resolve_port() -> tuple[int, str]:
    raw = os.environ.get("ERRORTA_SIDECAR_PORT")
    if raw:
        try:
            return int(raw), "env"
        except ValueError:
            pass
    return 8770, "default"


def _probe_ollama(host: str) -> OllamaHealth:
    try:
        with httpx.Client(timeout=1.5) as client:
            r = client.get(f"{host.rstrip('/')}/api/version")
            r.raise_for_status()
            body = r.json() if r.headers.get("content-type", "").startswith("application/json") else {}
            return OllamaHealth(host=host, reachable=True, version=body.get("version"))
    except Exception as e:  # noqa: BLE001
        return OllamaHealth(host=host, reachable=False, error=str(e)[:200])


def _processes_payload() -> list[ProcessPayload]:
    return [ProcessPayload(**p.to_dict()) for p in shell_processes.list_managed()]


# --- Endpoints --------------------------------------------------------------


@router.get("/status", response_model=StatusResponse)
def status() -> StatusResponse:
    procs = _processes_payload()
    sidecar = SidecarHealth(
        pid=os.getpid(),
        uptime_seconds=round(shell_processes.uptime_seconds(), 3),
        cold_start_seconds=shell_config.cold_start_seconds(),
    )
    ollama = _probe_ollama(shell_config.get_ollama_host())
    return StatusResponse(sidecar=sidecar, ollama=ollama, processes=procs)


@router.get("/processes", response_model=ProcessesResponse)
def processes() -> ProcessesResponse:
    return ProcessesResponse(processes=_processes_payload())


@router.get("/sidecar/port", response_model=SidecarPortResponse)
def sidecar_port() -> SidecarPortResponse:
    port, source = _resolve_port()
    return SidecarPortResponse(port=port, source=source)


@router.get("/config/ollama-host", response_model=OllamaHostResponse)
def get_ollama_host() -> OllamaHostResponse:
    return OllamaHostResponse(host=shell_config.get_ollama_host())


@router.post("/config/ollama-host", response_model=OllamaHostResponse)
def set_ollama_host(body: OllamaHostBody) -> OllamaHostResponse:
    try:
        host = shell_config.set_ollama_host(body.host)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    return OllamaHostResponse(host=host)


@router.post("/ready", response_model=ColdStartResponse)
def mark_ready() -> ColdStartResponse:
    shell_config.mark_ready()
    return ColdStartResponse(
        cold_start_seconds=shell_config.cold_start_seconds(),
        process_start_epoch=shell_config.process_start(),
    )


@router.get("/cold-start", response_model=ColdStartResponse)
def cold_start() -> ColdStartResponse:
    return ColdStartResponse(
        cold_start_seconds=shell_config.cold_start_seconds(),
        process_start_epoch=shell_config.process_start(),
    )
